typegen/strats/inline.py

```diff
@@ -1,96 +1,110 @@
-import ast
 import functools
 import logging
 import operator
 import pathlib
-import itertools
+
+import pandas as pd
+
+import libcst as cst
+
+from libcst.metadata import PositionProvider
 
 from constants import TraceData
 from tracing.trace_data_category import TraceDataCategory
 from typegen.strats.gen import TypeHintGenerator
 
-import pandas as pd
-
 logger = logging.getLogger(__name__)
 
 
-class TypeHintTransformer(ast.NodeTransformer):
-    def __init__(self, relevant: pd.DataFrame) -> None:
-        super().__init__()
-        self.df = relevant
+class TargetExtractor(cst.CSTVisitor):
+    targets: list[tuple[str, cst.Name | cst.Attribute]]
 
-    def generic_visit(self, node: ast.AST) -> ast.AST:
-        # Track ClassDefs and FunctionDefs to disambiguate
-        # functions from methods
-        if isinstance(node, ast.ClassDef):
-            for child in filter(lambda c: isinstance(c, ast.FunctionDef), node.body):
-                child.parent = node  # type: ignore
+    def __init__(self):
+        self.targets = list()
 
-        # Track assignments from Functions
-        # NOTE: this handles nested functions too, because the parent reference gets overwritten
-        # NOTE: before we start generating hints for its children
-        if isinstance(node, ast.FunctionDef):
-            for direct in ast.iter_child_nodes(node):
-                for child in ast.walk(direct):  # type: ignore
-                    child.parent = node  # type: ignore
+    def visit_Attribute(self, node: cst.Attribute) -> bool | None:
+        self.targets.append((node.attr.value, node))
+        return False
 
-        return super().generic_visit(node)
+    def visit_Name(self, node: cst.Name) -> bool | None:
+        self.targets.append((node.value, node))
+        return False
 
-    def visit_FunctionDef(self, fdef: ast.FunctionDef):
-        logger.debug(f"Applying hints to '{fdef.name}'")
 
-        self._add_parameter_hints(fdef)
-        self._add_return_hint(fdef)
+class TypeHintTransformer(cst.CSTTransformer):
+    METADATA_DEPENDENCIES = (PositionProvider,)
 
-        self.generic_visit(fdef)
-        return fdef
+    def __init__(self, relevant: pd.DataFrame) -> None:
+        super().__init__()
+        self.df = relevant
+        self._parent_stack: list[cst.FunctionDef | cst.ClassDef] = []
 
-    def _add_parameter_hints(self, fdef: ast.FunctionDef) -> None:
-        # parameters
-        param_masks = [
-            self.df[TraceData.CATEGORY] == TraceDataCategory.FUNCTION_PARAMETER,
-            self.df[TraceData.FUNCNAME] == fdef.name,
-        ]
-        params = self.df[functools.reduce(operator.and_, param_masks)]
+    def _outermost_class(self) -> cst.ClassDef | None:
```

Comment:
Shouldn't it be the inner most class? It gets the most inner class, but the naming says otherwise.

Reply:
True